### PR TITLE
Implement templates

### PR DESCRIPTION
**Pull request by NCPlayz**
### Summary

This implements the fetching of templates through their codes, and the creation of guilds through templates. The endpoint `/guilds/:id/templates` is off bounds for bots. It also provides basic stateless access (as well as methods) from the data received through fetching the template. It also adds an optional kwarg to `Client.create_guild`, `code`, which allows you to provide a template code and create a guild from it rather than fetching the template first.
I still feel like there is definitely room for improvement so for now it will be a draft PR.

### Checklist

- [x] If code changes were made then they have been tested.
    - [x] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Comment by apple502j** [comment]
Will `DELETE /guilds/templates/code` be implemented? (for deleting template)

**Comment by NCPlayz** [comment]
Deleting templates follows `DELETE /guilds/:id/templates`, and that is attached to the guild, not under `/guilds/templates/:code`. All methods for `/guilds/:id/templates` don't seem to be available to bots as of now.

File: discord/template.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+        id = _get_as_snowflake(data, 'source_guild_id')
+        source_serialised = data['serialized_source_guild']
+        source_serialised['id'] = id
+        state = _PartialTemplateState(state=self._state)
+
+        self.source_guild = Guild(data=source_serialised, state=state)
+
+    def __repr__(self):
+        return '<Template code={0.code!r} uses={0.uses} name={0.name!r} creator={0.creator!r} source_guild={0.source_guild!r}>'.format(self)
```
Comment (Rapptz) [inline discord/template.py:95]: This line is a bit too long. Split it a bit.

File: discord/template.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        The creator of the template.
+    created_at: :class:`datetime.datetime`
+        When the template was created.
+    updated_at: :class:`datetime.datetime`
+        When the template was last updated (referred to as "last synced" in the client).
+    source_guild: :class:`TemplateGuild`
+        The source guild.
+    """
+
+    def __init__(self, *, state, data, custom=False):
```
Comment (Rapptz) [inline discord/template.py:74]: What is this `custom` parameter doing? It's unused.

File: discord/utils.py [hunk 3]
```diff
@@ -436,6 +436,17 @@ def resolve_invite(invite):
             return m.group(1)
     return invite
 
+def resolve_template(code):
+    from .template import Template # circular import
+    if isinstance(code, (Template, Object)):
+        return template.id
+    else:
+        rx = r'(?:https?\:\/\/)?discord(?:\.new|(?:app)?\.com\/templates)\/(.+)'
```
Comment (Rapptz) [inline discord/utils.py:444]: This regex seems wrong. `discordapp.com/templates/...` doesn't seem to actually be valid (nor discord.com/templates/...).

```suggestion
        rx = r'(?:https?\:\/\/)?discord\.new\/(.+)'
```

Comment (mdashlw) [reply discord/utils.py]: It's "template" not "templates"
image

File: discord/client.py [hunk 4]
```diff
@@ -1019,6 +1020,11 @@ def fetch_guilds(self, *, limit=100, before=None, after=None):
         """
         return GuildIterator(self, limit=limit, before=before, after=after)
 
+    async def fetch_template(self, code):
```
Comment (Rapptz) [inline discord/client.py:1023]: Needs documentation.

File: discord/client.py [hunk 5]
```diff
@@ -1070,6 +1076,8 @@ async def create_guild(self, name, region=None, icon=None):
         icon: :class:`bytes`
             The :term:`py:bytes-like object` representing the icon. See :meth:`.ClientUser.edit`
             for more details on what is expected.
+        code: Optional[:class:`str`]
+            The code for a template to create the guild with.
```
Comment (Rapptz) [inline discord/client.py:1080]: Needs versionadded string

File: discord/template.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        return None
+
+    async def query_members(self, **kwargs):
+        return []
+
+    def __getattr__(self, attr):
+        raise AttributeError('PartialTemplateState does not support {0!r}.'.format(attr))
+
+class Template:
+    """Represents a Discord template.
```
Comment (Rapptz) [inline discord/template.py:52]: Needs versionadded

**Pull request merged**